pricegraph/src/api/price_estimation.rs

```diff
@@ -26,30 +24,60 @@ impl Pricegraph {
             sell: pair.buy,
         };
 
-        let mut last_exchange_rate = None;
-        if sell_amount <= 0.0 {
+        if sell_amount == 0.0 {
             // NOTE: For a 0 volume we simulate sending an tiny epsilon of value
             // through the network without actually filling any orders.
+            let mut exchange_rate = None;
             let flow = orderbook.fill_optimal_transitive_order_if(inverse_pair, |flow| {
-                last_exchange_rate = Some(flow.exchange_rate);
+                exchange_rate = Some(flow.exchange_rate);
                 false
             });
             debug_assert!(flow.is_none());
+
+            // NOTE: The exchange rates are for transitive orders in the inverse
+            // direction, so we need to invert the exchange rate and account for
+            // the fees so that the estimated exchange rate actually overlaps with
+            // the last counter transtive order's exchange rate.
+            return Some(exchange_rate?.inverse().price().value());
+        }
+
+        if !num::is_strictly_positive_and_finite(sell_amount) {
+            return None;
         }
 
-        let mut remaining_volume = sell_amount;
-        while remaining_volume > 0.0 {
-            let flow = orderbook.fill_optimal_transitive_order(inverse_pair)?;
-            last_exchange_rate = Some(flow.exchange_rate);
-            remaining_volume -= flow.capacity;
+        // NOTE: Iteratively compute the how much total buy volume (liquidity)
+        // is available at successively "worse" exchange rates until all the
+        // specified sell amount can be used to buy the available liquidity at
+        // the marginal exchange rate.
+        let mut total_buy_volume = 0.0;
+        let mut maximum_sell_volume = 0.0;
+        while let Some(flow) = orderbook.fill_optimal_transitive_order_if(inverse_pair, |flow| {
+            let current_exchange_rate = match ExchangeRate::new(total_buy_volume / sell_amount) {
```

Comment:
nit: the naming `sell_amount` & `maximum_sell_amount` was confusing me here. It's important that this amount is the input into the function and not the `sell_amount` so far. Maybe we can rename `total_buy_amount` and `maximum_sell_amount` to something that reflects that (`xxx_amount_so_far` for the lack of a better suggestion).

Reply:
How about `cumulative_{buy,sell}_amount`?

Reply:
I like you suggestion. I would still name sell_amount max_sell_amount to make it more obvious that the `current_exchange_rate` is not computed between the cumulative amounts.